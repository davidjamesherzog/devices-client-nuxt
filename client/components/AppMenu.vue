<template>
  <b-menu class="app-menu section">
    <b-menu-list label="General">
      <div v-for="(item, key) of items" :key="key" @click="close()">
        <b-menu-item
          :label="item.title"
          :icon="item.icon"
          tag="router-link"
          :to="item.route"
          :data-menu="item.title"
          :class="{ 'has-text-weight-bold': isCurrentRoute(item.route) }"
        ></b-menu-item>
      </div>
    </b-menu-list>
  </b-menu>
</template>

<script lang="ts">
import { Component, Vue } from 'nuxt-property-decorator';
import Link from '~/client/models/Link';
import Route from '~/client/models/Route';

@Component({
  name: 'AppMenu'
})
export default class Menu extends Vue {
  private items: Array<Link> = [
    {
      title: 'Home',
      icon: 'home',
      route: { name: 'index' }
    },
    {
      title: 'Devices',
      icon: 'devices',
      route: { name: 'devices' }
    },
    {
      title: 'Networks',
      icon: 'lan',
      route: { name: 'networks' }
    }
  ];

  public isCurrentRoute(route: Route): boolean {
    return this.$route.name === route.name;
  }

  public close(): void {
    this.$nuxt.$emit('close-menu');
  }
}
</script>

<style lang="scss" scoped>
.p-1 {
  padding: 1em;
}
</style>
