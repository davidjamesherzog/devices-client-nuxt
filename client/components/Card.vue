<template>
  <div class="column">
    <div class="card">
      <header class="card-header">
        <p class="card-header-title has-text-grey" data-card-title>
          {{ title }}
        </p>
      </header>
      <div class="card-content">
        <div class="content has-text-centered">
          <b-icon
            :icon="icon"
            size="is-large"
            type="is-primary"
            data-card-icon
          />
        </div>
      </div>
      <footer class="card-footer">
        <div class="card-footer-item">
          <span>
            <slot />
          </span>
        </div>
      </footer>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'nuxt-property-decorator';

@Component({
  name: 'Card'
})
export default class Card extends Vue {
  @Prop({ type: String })
  private title!: string;

  @Prop({ type: String })
  private icon!: string;
}
</script>
