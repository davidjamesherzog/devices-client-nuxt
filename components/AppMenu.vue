<template>
  <b-menu class="app-menu section">
    <b-menu-list label="General">
      <div v-for="(item, key) of items" :key="key" @click="close()">
        <b-menu-item
          :label="item.title"
          :icon="item.icon"
          tag="router-link"
          :to="item.to"
          :data-menu="item.title"
        ></b-menu-item>
      </div>
    </b-menu-list>
  </b-menu>
</template>

<script lang="ts">
import { Component, Vue } from 'nuxt-property-decorator';

@Component({
  name: 'AppMenu'
})
export default class Menu extends Vue {
  private items: Array<any> = [
    {
      title: 'Home',
      icon: 'home',
      to: { name: 'index' }
    },
    {
      title: 'Devices',
      icon: 'devices',
      to: { name: 'devices' }
    },
    {
      title: 'Networks',
      icon: 'lan',
      to: { name: 'networks' }
    }
  ];

  public close(): void {
    this.$nuxt.$emit('close-menu');
  }
}
</script>

<style lang="scss" scoped>
.p-1 {
  padding: 1em;
}
</style>
