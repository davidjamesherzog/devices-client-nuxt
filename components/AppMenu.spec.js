import { createLocalVue, shallowMount } from '@vue/test-utils';
import Buefy from 'buefy';
import AppMenu from '~/components/AppMenu.vue';
import Card from '~/components/Card';

describe('AppMenu', () => {
  describe('close', () => {
    test('should emit event', () => {
      const appMenu = new AppMenu();
      appMenu.$nuxt = {
        $emit: jest.fn()
      };
      appMenu.close();
      expect(appMenu.$nuxt.$emit).toHaveBeenCalledWith('close-menu');
    });
  });

  describe('template', () => {
    let localVue;

    beforeEach(() => {
      localVue = createLocalVue();
      localVue.use(Buefy);
    });

    test('is a Vue instance', () => {
      const wrapper = shallowMount(AppMenu, { localVue });
      expect(wrapper.vm).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    test('should render home link', () => {
      const wrapper = shallowMount(AppMenu, {
        localVue
      });
      const title = wrapper.find('[data-menu="Home"]');
      expect(title.exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    test('should render devices link', () => {
      const wrapper = shallowMount(AppMenu, {
        localVue
      });
      const title = wrapper.find('[data-menu="Devices"]');
      expect(title.exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    test('should render networks link', () => {
      const wrapper = shallowMount(AppMenu, {
        localVue
      });
      const title = wrapper.find('[data-menu="Networks"]');
      expect(title.exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
