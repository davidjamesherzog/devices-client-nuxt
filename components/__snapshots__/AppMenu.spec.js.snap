// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppMenu template is a Vue instance 1`] = `
<b-menu-stub accordion="true" activable="true" class="app-menu section">
  <b-menu-list-stub label="General" ariarole="" size="is-small">
    <div>
      <b-menu-item-stub label="Home" icon="home" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Home"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Devices" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Devices"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Networks" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Networks"></b-menu-item-stub>
    </div>
  </b-menu-list-stub>
</b-menu-stub>
`;

exports[`AppMenu template should render devices link 1`] = `
<b-menu-stub accordion="true" activable="true" class="app-menu section">
  <b-menu-list-stub label="General" ariarole="" size="is-small">
    <div>
      <b-menu-item-stub label="Home" icon="home" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Home"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Devices" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Devices"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Networks" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Networks"></b-menu-item-stub>
    </div>
  </b-menu-list-stub>
</b-menu-stub>
`;

exports[`AppMenu template should render home link 1`] = `
<b-menu-stub accordion="true" activable="true" class="app-menu section">
  <b-menu-list-stub label="General" ariarole="" size="is-small">
    <div>
      <b-menu-item-stub label="Home" icon="home" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Home"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Devices" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Devices"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Networks" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Networks"></b-menu-item-stub>
    </div>
  </b-menu-list-stub>
</b-menu-stub>
`;

exports[`AppMenu template should render networks link 1`] = `
<b-menu-stub accordion="true" activable="true" class="app-menu section">
  <b-menu-list-stub label="General" ariarole="" size="is-small">
    <div>
      <b-menu-item-stub label="Home" icon="home" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Home"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Devices" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Devices"></b-menu-item-stub>
    </div>
    <div>
      <b-menu-item-stub label="Networks" icon="lightbulb" animation="slide" tag="router-link" ariarole="" size="is-small" to="[object Object]" data-menu="Networks"></b-menu-item-stub>
    </div>
  </b-menu-list-stub>
</b-menu-stub>
`;
